File: python/src/BFS_Graphs.py
from collections import deque


def search(name):
    search_queue = deque()
    search_queue += graph[name]
    searched = []
    while search_queue:
        person = search_queue.popleft()
        if not person in searched:
            if person_is_seller(person):
                print(person + " is a mango seller!")
                return True
            else:
                search_queue += graph[person]
                searched.append(person)
    return False


def person_is_seller(name):
    return name[-1] == 'm'


graph = {"you": ["alice", "bob", "claire"], "bob": ["anuj", "peggy"], "alice": ["peggy"], "claire": ["thom", "jonny"],
         "anuj": [], "peggy": [], "thom": [], "jonny": []}

search("you")

File: java/src/BFSGraphs.java
import java.util.*;

public class BFSGraphs {

    private static final Map<String, List<String>> graph = new HashMap<>();

    static {
        graph.put("you", List.of("alice", "bob", "claire"));
        graph.put("bob", List.of("anuj", "peggy"));
        graph.put("alice", List.of("peggy"));
        graph.put("claire", List.of("thom", "jonny"));
        graph.put("anuj", List.of());
        graph.put("peggy", List.of());
        graph.put("thom", List.of());
        graph.put("jonny", List.of());
    }

    private static boolean personIsSeller(String name) {
        return name.endsWith("m");
    }

    public static String search(String name) {
        var queue = new LinkedList<>(graph.get(name));
        var searched = new ArrayList<String>();
        while (!queue.isEmpty()) {
            var person = queue.pop();
            if (searched.contains(person)) {
                continue;
            }
            if (personIsSeller(person)) {
                return person + " is a mango seller!";
            } else {
                searched.add(person);
                queue.addAll(graph.get(person));
            }
        }
        return null;
    }

    public static void main(String[] args) {
        System.out.println(search("you"));
    }
}
